### PR TITLE
Added functionality to determine polar second moment of area and section modulus of a polygon

### PR DESCRIPTION
**Pull request by ishanaj**
[Polar modulus](https://en.wikipedia.org/wiki/Polar_moment_of_inertia) and [section modulus](https://en.wikipedia.org/wiki/Section_modulus) are properties of a polygon (more specifically a cross-section).
 
With `beam` module now accepting the cross-section (geometry object) as an alternative to the `second_moment`, this functionality would be of importance to the `beam` module, as it will be able to calculate the bending stresses and shear stresses on a beam with a particular cross-section
This is in addition to previously implemented `second_moment_of_area()`. Also it takes a parameter `point` similar to that in the `second_moment_of_area()`

ToDo's:
- [x] Tests
- [x] Documentation
- [x] Implementing the same for Ellipses class
- [x] method to calculate first moment of area (for polygon)

#### Release Notes

- geometry
    - added methods in the polygon class to calculate section modulus, polar second moment of area and first moment of area of a two-dimensional polygon
    - added methods in the ellipse class to calculate section modulus and polar second moment of area of a two-dimensional ellipse

### CODE REVIEW

File: sympy/geometry/polygon.py [hunk 1]
```diff
@@ -438,6 +438,32 @@ def second_moment_of_area(self, point=None):
...
 
 
+    def polar_modulus(self):
+        """Returns the polar modulus of a two-dimensional polygon"""
+        return self.second_moment_of_area()[0] + self.second_moment_of_area()[1]
+
+
+    def section_modulus(self, point=None):
+        """Returns a tuple with the section modulus of a two-dimensional
+        polygon.
```
Comment (jashan498) [inline sympy/geometry/polygon.py]: You should add the description of the `point` argument in docstring and also that its default value is the point farthest from the neutral axis.

**Comment by smichr** [comment]
👓 I did not notice the ability to make a draft PR before...good idea.

File: sympy/geometry/polygon.py [hunk 2]
```diff
@@ -438,6 +438,32 @@ def second_moment_of_area(self, point=None):
...
+        """
+        x_c, y_c = self.centroid.x, self.centroid.y
+        if point is None:
+            # taking x and y as maximum distances from centroid
+            x_min, y_min, x_max, y_max = self.shape.bounds
```
Comment (smichr) [inline sympy/geometry/polygon.py]: I am not aware of a `shape` attribute. Do you just mean `self.bounds`?

Comment (ishanaj) [reply sympy/geometry/polygon.py]: Yes, it is actually `shape.bounds`. I mistakenly got this from other file

File: sympy/geometry/polygon.py [hunk 3]
```diff
@@ -438,6 +438,70 @@ def second_moment_of_area(self, point=None):
...
+
+        Examples
+        ========
+
+        >>> from sympy import symbols, Polygon, Point
+        >>> a, b = symbols('a, b', positive=True)
+        >>> rectangle = Polygon((0, 0), (a, 0), (a, b), (0, b))
+        >>> rectangle.section_modulus()
+        (a*b**2/6, a**2*b/6)
+        >>> rectangle.section_modulus(Point(a/4, b/4))
```
Comment (jashan498) [inline sympy/geometry/polygon.py]: Shouldn't we return  a `ValueError` if  the `Point` lies outside of the `Polygon`?
```
>>> rectangle = Polygon((0, 0), (a, 0), (a, b), (0, b))
>>> rectangle.section_modulus(Point(10*a, 10*b))
(a*b**2/114, a**2*b/114)
```

Comment (ishanaj) [reply sympy/geometry/polygon.py]: But I guess then it won't be able to handle arbitrary points like:
 `rectangle.section_modulus(Point(x, y))`
 I have added such an example in the tests.

**Comment by ishanaj** [comment]
ping @moorepants @oscarbenjamin

File: sympy/geometry/polygon.py [hunk 4]
```diff
@@ -438,6 +438,70 @@ def second_moment_of_area(self, point=None):
...
+        Examples
+        ========
+
+        >>> from sympy import Polygon, symbols
+        >>> a, b = symbols('a, b')
+        >>> rectangle = Polygon((0, 0), (a, 0), (a, b), (0, b))
+        >>> rectangle.polar_modulus()
+        a**3*b/12 + a*b**3/12
+        """
+        return self.second_moment_of_area()[0] + self.second_moment_of_area()[1]
```
Comment (oscarbenjamin) [inline sympy/geometry/polygon.py]: Can we do this by only calling second_moment_of_area once?

File: sympy/geometry/polygon.py [hunk 5]
```diff
@@ -438,6 +438,70 @@ def second_moment_of_area(self, point=None):
...
+
+        >>> from sympy import symbols, Polygon, Point
+        >>> a, b = symbols('a, b', positive=True)
+        >>> rectangle = Polygon((0, 0), (a, 0), (a, b), (0, b))
+        >>> rectangle.section_modulus()
+        (a*b**2/6, a**2*b/6)
+        >>> rectangle.section_modulus(Point(a/4, b/4))
+        (-a*b**2/3, -a**2*b/3)
+        """
+        x_c, y_c = self.centroid.x, self.centroid.y
```
Comment (oscarbenjamin) [inline sympy/geometry/polygon.py]: Presumably self.centroid can be unpacked directly:
```
x, y = self.centroid
```

File: sympy/geometry/polygon.py [hunk 6]
```diff
@@ -438,6 +438,70 @@ def second_moment_of_area(self, point=None):
...
+            x_min, y_min, x_max, y_max = self.bounds
+            y = max(y_c - y_min, y_max - y_c)
+            x = max(x_c - x_min, x_max - x_c)
+        else:
+            # taking x and y as distances of the given point from the centroid
+            y = point.y - y_c
+            x = point.x - x_c
+
+        S_x =  self.second_moment_of_area()[0]/y
+        S_y = self.second_moment_of_area()[1]/x
```
Comment (oscarbenjamin) [inline sympy/geometry/polygon.py]: Spacing is inconsistent (also calling second_moment twice)

File: sympy/geometry/polygon.py [hunk 7]
```diff
@@ -438,6 +438,70 @@ def second_moment_of_area(self, point=None):
...
+        >>> a, b = symbols('a, b', positive=True)
+        >>> rectangle = Polygon((0, 0), (a, 0), (a, b), (0, b))
+        >>> rectangle.section_modulus()
+        (a*b**2/6, a**2*b/6)
+        >>> rectangle.section_modulus(Point(a/4, b/4))
+        (-a*b**2/3, -a**2*b/3)
+        """
+        x_c, y_c = self.centroid.x, self.centroid.y
+        if point is None:
+            # taking x and y as maximum distances from centroid
```
Comment (oscarbenjamin) [inline sympy/geometry/polygon.py]: Is it useful to have this default?

**Comment by oscarbenjamin** [comment]
I'm not familiar with these concepts but it looks like e.g. the section modulus is defined relative to the neutral axis rather than to a particular point. Maybe it makes more sense to pass the argument as a line rather than a point...

Also is the intention to add these features to Ellipse etc as well?

**Comment by ishanaj** [comment]
I think if we accept a line instead of a point we might have to check it whether it is parallel to the centroidal axis.
Also then we might need to accept two lines for calculating it about x and y axis.
But if we take a point, we simply assume two perpendicular lines passing through it and parallel to the centroidal axis, and can calculate the section modulus about both x and y axis.

I guess this is also the reason why even `second_moment_of area()` takes a point as a parameter.
(And of course I have taken the idea from it).

Yes, if this is ok we can add similar function to Ellipses.

File: sympy/geometry/polygon.py (hunk 4 above)
Comment (ishanaj) [reply sympy/geometry/polygon.py]: yes, I will be changing it

**Comment by ishanaj** [comment]
ping @moorepants

File: sympy/geometry/ellipse.py [hunk 8]
```diff
@@ -1385,6 +1386,75 @@ def second_moment_of_area(self, point=None):
...
+
+        >>> from sympy import Symbol, Ellipse, Circle, Point2D
+        >>> d = Symbol('d', positive=True)
+        >>> c = Circle((0, 0), d/2)
+        >>> c.section_modulus()
+        (pi*d**3/32, pi*d**3/32)
+        >>> e = Ellipse(Point2D(0, 0), 2, 4)
+        >>> e.section_modulus()
+        (8*pi, 4*pi)
+        """
```
Comment (moorepants) [inline sympy/geometry/ellipse.py]: Please add some references that point to explanations of a "section modulus".

Comment (moorepants) [reply sympy/geometry/ellipse.py]: I don't know what a "section modulus" is based on this documentation.

**Comment by moorepants** [comment]
It would be helpful to show examples and references that indicate what a polar modulus and section modulus are. The unit tests should include tests that verify textbook problems work as excepted for a variety of supported shapes. There is no accompanying prose style sphinx documentation to explain these concepts and how they are useful. I'm not even sure how this will be used with the Beam class.

**Comment by jashan498** [comment]
ping @ishanaj

**Comment by ishanaj** [comment]
Since PR #17001 has been merged I will be further adding a method to calculate first moment too and would surely improve the documentation considering Jason's last comment.

**Comment by ishanaj** [comment]
Will be adding some tests next.

**Comment by ishanaj** [comment]
@jashan498 I have made the required changes. I think the PR is ready to be merged once the tests pass.

**Review by jashan498 — APPROVED** [review]
Looks good to me

**Comment by ishanaj** [comment]
Thanks for merging!

**Pull request merged**